src/appleseed-max-impl/appleseedblendmtl/appleseedblendmtl.cpp

```diff
@@ -0,0 +1,744 @@
+        if (result != IO_OK)
+            break;
+
+        switch (iload->CurChunkID())
+        {
+          case ChunkFileFormatVersion:
+            {
+                USHORT version;
+                result = read<USHORT>(iload, &version);
+            }
+            break;
+
+          case ChunkMtlBase:
+            result = MtlBase::Load(iload);
+            break;
+        }
+
+        if (result != IO_OK)
+            break;
+
+        result = iload->CloseChunk();
+        if (result != IO_OK)
+            break;
+    }
+
+    return result;
+}
+
+Color AppleseedBlendMtl::GetAmbient(int mtlNum, BOOL backFace)
+{
+    return Color(0.0f, 0.0f, 0.0f);
+}
+
+Color AppleseedBlendMtl::GetDiffuse(int mtlNum, BOOL backFace)
+{
+    return Color(0.0f, 0.0f, 0.0f);
+}
+
+Color AppleseedBlendMtl::GetSpecular(int mtlNum, BOOL backFace)
+{
+    return Color(0.0f, 0.0f, 0.0f);
+}
+
+float AppleseedBlendMtl::GetShininess(int mtlNum, BOOL backFace)
+{
+    return 0.0f;
+}
+
+float AppleseedBlendMtl::GetShinStr(int mtlNum, BOOL backFace)
+{
+    return 0.0f;
+}
+
+float AppleseedBlendMtl::GetXParency(int mtlNum, BOOL backFace)
+{
+    return 0.0f;
+}
+
+void AppleseedBlendMtl::SetAmbient(Color c, TimeValue t)
+{
+}
+
+void AppleseedBlendMtl::SetDiffuse(Color c, TimeValue t)
+{
+}
+
+void AppleseedBlendMtl::SetSpecular(Color c, TimeValue t)
+{
+}
+
+void AppleseedBlendMtl::SetShininess(float v, TimeValue t)
+{
+}
+
+void AppleseedBlendMtl::Shade(ShadeContext& sc)
+{
+}
+
+int AppleseedBlendMtl::NumSubMtls()
+{
+    return MtlCount; // MtlCount;
+}
+
+Mtl* AppleseedBlendMtl::GetSubMtl(int i)
+{
+    Mtl* mtl = nullptr;
+    if (i==0)
+        m_pblock->GetValue(ParamIdBaseMtl, 0, mtl, FOREVER);
+    else
+        m_pblock->GetValue(ParamIdLayerMtl, 0, mtl, FOREVER, i-1);
+    return mtl;
+}
+
+void AppleseedBlendMtl::SetSubMtl(int i, Mtl* m)
+{
+    if (i == 0)
+        m_pblock->SetValue(ParamIdBaseMtl, 0, m);
+    else
+        m_pblock->SetValue(ParamIdLayerMtl, 0, m, i-1);
+}
+
+MSTR AppleseedBlendMtl::GetSubMtlSlotName(int i)
+{
+    const auto mtl_id = static_cast<MtlId>(i);
+    return g_material_slot_names[mtl_id];
+}
+
+int AppleseedBlendMtl::get_sides() const
+{
+    return asr::ObjectInstance::BothSides;
+}
+
+bool AppleseedBlendMtl::can_emit_light() const
+{
+    return false;
+}
+
+asf::auto_release_ptr<asr::Material> AppleseedBlendMtl::create_material(
+    asr::Assembly&  assembly,
+    const char*     name,
+    const bool      use_max_procedural_maps)
+{
+    return
+        use_max_procedural_maps
+            ? create_builtin_material(assembly, name)
+            : create_osl_material(assembly, name);
+}
+
+asf::auto_release_ptr<asr::Material> AppleseedBlendMtl::create_osl_material(
+    asr::Assembly&  assembly,
+    const char*     name)
+{
+    Mtl* mat = nullptr;
+    m_pblock->GetValue(ParamIdBaseMtl, 0, mat, FOREVER);
+
+    if (!mat)
+        return asr::OSLMaterialFactory().create(name, asr::ParamArray());
```

Comment:
What happens in this case?

Reply:
It just returns default osl material, which is just black. I couldn't think about any other way to cancel the creation of material.

Reply:
Actually I can create that material in the beginning and then supply parameters. Or return it with default parameters.

Reply:
If I can access parameters later...

Reply:
You can always access the parameters of an entity using `entity->get_parameters()` (which returns a mutable `renderer::ParamArray` so you can add/modify entries).